docs/blog-2/blog-2.md

```diff
@@ -0,0 +1,194 @@
+Also, the inference doesn't happen for any other `show`/`go` direction (like `down` or `next`) since that behavior could be pretty confusing and generally of little practical use.
+
+### `update`
+
+Let's check out `drop-constraint` back again and, without changing anything in `.git/machete`, fire `git machete update`:
+
+![git machete update](update-3.png)
+
+`update` couldn't find an entry for `drop-constraint` in the definition file, so it fell back to automatic inference.
+
+### `add`
+
+Now let's try and do `git machete add`:
+
+![git machete add](add.png)
+
+Since the desired upstream branch wasn't specified (no `--onto` option was provided), the `add` subcommand inferred the `drop-constraint`'s upstream.
+Note that in the case of both `update` and `add` the user has been explicitly asked whether git machete's guess is acceptable.
+
+### `discover`
+
+What's more, the inference is not just limited to a single branch - it can even be performed on a repository where there is not yet a `.git/machete` file to discover the entire dependency tree with a single command!
+
+For demonstration purposes, let's now remove the `.git/machete` file (so as to make sure `git machete` isn't provided with any hint) and run `git machete discover`:
+
+![git machete discover](discover.png)
+
+`discover` gives the choice to either accept the discovered tree right away with `y[es]`, `e[dit]` it first, or reject it with `n[o]`.
+In the case of `yes`/`edit`, the old definition file (if it already exists) will be saved under `.git/machete~` (note the added tilde).
+
+Under the hood, `discover` simply performed upstream inference (just as for `show up` etc.) for every single local branch independently,
+while applying some tricks (inspired by the [disjoint-set data structure](https://en.wikipedia.org/wiki/Disjoint-set_data_structure)) to make sure no cycles appear in the created graph.
+
+The only thing that obviously could not be inferred were custom annotations.
+Also, if the tree structure has been changed in the meantime (so that e.g. `develop` became a child of `call-ws` instead of the inverse), `discover` can't be expected to exactly guess the changed structure.
+The inference is based on git reflogs and doesn't know anything about the current or previous state of the `.git/machete` file.
+
+At this point one can ask a question: why then is the definition file even needed since the upstreams could be always inferred on the fly?
+The argument against such an approach is that it wouldn't be desirable if everything in the tool happened automagically/behind the scenes.
+The developer should keep a sensible amount of control while still being provided with some hints when they're in doubt.
+It seems much more reasonable than the inverse - inferring everything that's possible and only asking the human when the tool is in doubt.
```

Comment:
Nit: pls replace `-` (hyphen) with `&mdash;` where appropriate &mdash; see https://virtuslab.github.io/tips/#english/grammar-tips-2

Reply:
Basically, all occurrences of ` - ` (space-hyphen-space) can/should probably be replaced with an em dash